README.md
[![Build Status](https://travis-ci.org/Kage/Mojolicious-Plugin-TrustedProxy.svg?branch=master)](https://travis-ci.org/Kage/Mojolicious-Plugin-TrustedProxy)
[![Coverage Status](https://coveralls.io/repos/Kage/Mojolicious-Plugin-TrustedProxy/badge.svg?branch=master)](https://coveralls.io/r/Kage/Mojolicious-Plugin-TrustedProxy?branch=master)
[![Kwalitee status](http://cpants.cpanauthors.org/dist/Mojolicious-Plugin-TrustedProxy.png)](http://cpants.charsbar.org/dist/overview/Mojolicious-Plugin-TrustedProxy)
[![Cpan license](https://img.shields.io/cpan/l/Mojolicious-Plugin-TrustedProxy.svg)](https://metacpan.org/release/Mojolicious-Plugin-TrustedProxy)
[![Cpan version](https://img.shields.io/cpan/v/Mojolicious-Plugin-TrustedProxy.svg)](https://metacpan.org/release/Mojolicious-Plugin-TrustedProxy)
[![GitHub tag](https://img.shields.io/github/tag/Kage/Mojolicious-Plugin-TrustedProxy.svg)]()
[![GitHub issues](https://img.shields.io/github/issues/Kage/Mojolicious-Plugin-TrustedProxy.svg)](https://github.com/Kage/Mojolicious-Plugin-TrustedProxy/issues)

# NAME

Mojolicious::Plugin::TrustedProxy - Mojolicious plugin to set the remote
address, connection scheme, and more from trusted upstream proxies

# VERSION

Version 0.05

# SYNOPSIS

    use Mojolicious::Lite;

    plugin 'TrustedProxy' => {
      enabled         => 1,
      ip_headers      => ['x-forwarded-for', 'x-real-ip'],
      scheme_headers  => ['x-forwarded-proto', 'x-ssl'],
      https_values    => ['https', 'on', '1', 'true', 'enable', 'enabled'],
      parse_rfc7239   => 1,
      trusted_sources => ['127.0.0.0/8', '10.0.0.0/8'],
      hide_headers    => 0,
    };

    # Example of how you could verify expected functionality
    get '/test' => sub {
      my $c = shift;
      $c->render(json => {
        'tx.remote_address'            => $c->tx->remote_address,
        'tx.remote_proxy_address'      => $c->tx->remote_proxy_address,
        'req.url.base.scheme'          => $c->req->url->base->scheme,
        'req.url.base.host'            => $c->req->url->base->host,
        'is_trusted_source'            => $c->is_trusted_source,
        'is_trusted_source("1.1.1.1")' => $c->is_trusted_source('1.1.1.1'),
      });
    };

    app->start;

# DESCRIPTION

[Mojolicious::Plugin::TrustedProxy](https://metacpan.org/pod/Mojolicious::Plugin::TrustedProxy) modifies every [Mojolicious](https://metacpan.org/pod/Mojolicious) request
transaction to override connecting user agent values only when the request
comes from trusted upstream sources. You can specify multiple request headers
where trusted upstream sources define the real user agent IP address or the
real connection scheme, or disable either, and can hide the headers from the
rest of the application if needed.

This plugin provides much of the same functionality as setting
`MOJO_REVERSE_PROXY=1`, but with more granular control over what headers to
use and what upstream sources can send them. This is especially useful if your
Mojolicious app is directly exposed to the internet, or if it sits behind
multiple upstream proxies. You should therefore ensure your application does
not enable the default Mojolicious reverse proxy handler when using this plugin.

This plugin supports parsing [RFC 7239](http://tools.ietf.org/html/rfc7239)
compliant `Forwarded` headers, validates all IP addresses, and will
automatically convert RFC-4291 IPv4-to-IPv6 mapped values (useful for when your
Mojolicious listens on both IP versions). Please be aware that `Forwarded`
headers are only partially supported. More information is available in ["BUGS"](#bugs).

Debug logging can be enabled by setting the `MOJO_TRUSTEDPROXY_DEBUG`
environment variable. This plugin also adds a `remote_proxy_address`
attribute into `Mojo::Transaction`. If a remote IP address override header is
matched from a trusted upstream proxy, then `tx->remote_proxy_address`
will be set to the IP address of that proxy.

# CONFIG

## enabled

This allows you to load this plugin to access the [helper](#helpers) functions
without automatically running the `around_dispatch` hook on each request.
Default is `1` (enabled).

## ip\_headers

List of zero, one, or many HTTP headers where the real user agent IP address
will be defined by the trusted upstream sources. The first matched header is
used. An empty value will disable this and keep the original scheme value.
Default is `['x-forwarded-for', 'x-real-ip']`.

If a header is matched in the request, then `tx->remote_address` is set to
the value, and `tx->remote_proxy_address` is set to the IP address of the
upstream source.

## scheme\_headers

List of zero, one, or many HTTP headers where the real user agent connection
scheme will be defined by the trusted upstream sources. The first matched
header is used. An empty value will disable this and keep the original remote
address value. Default is `['x-forwarded-proto', 'x-ssl']`.

This tests that the header value is "truthy" but does not contain the literal
barewords `http`, `off`, or `false`. If the header contains any other
"truthy" value, then `req->url->base->scheme` is set to `https`.

## protocol\_headers

Alias for ["scheme\_headers"](#scheme_headers).

## https\_values

List of values to consider as "truthy" when evaluating the headers in
["scheme\_headers"](#scheme_headers). Default is
`['https', 'on', '1', 'true', 'enable', 'enabled']`.

## parse\_rfc7239

Enable support for parsing [RFC 7239](http://tools.ietf.org/html/rfc7239)
compliant `Forwarded` HTTP headers. Default is `1` (enabled).

If a `Forwarded` header is matched, the following actions occur with the first
semicolon-delimited group of parameters found in the header value:

- If the `for` parameter is found, then `tx->remote_address` is set to the
first matching value.
- If the `by` parameter is found, then `tx->remote_proxy_address` is set
to the first matching value, otherwise it is set to the IP address of the
upstream source.
- If the `proto` parameter is found, then `req->url->base->scheme` is set
to the first matching value.
- If the `host` parameter is found, then `req->url->base->host` is set to
the first matching value.

**Note!** If enabled, the headers defined in ["ip\_headers"](#ip_headers) and
["scheme\_headers"](#scheme_headers) will be overridden by any corresponding values found in
the `Forwarded` header.

## parse\_forwarded

Alias for ["parse\_rfc7239"](#parse_rfc7239).

## trusted\_sources

List of one or more IP addresses or CIDR classes that are trusted upstream
sources. (**Warning!** An empty value will trust from all IPv4 sources!) Default
is `['127.0.0.0/8', '10.0.0.0/8']`.

Supports all IP, CIDR, and range definition types from [Net::CIDR::Lite](https://metacpan.org/pod/Net::CIDR::Lite).

## hide\_headers

Hide all headers defined in ["ip\_headers"](#ip_headers), ["scheme\_headers"](#scheme_headers), and
`Forwarded` from the rest of the application when coming from trusted upstream
sources. Default is `0` (disabled).

# HELPERS

## is\_trusted\_source

    # From Controller context
    sub get_page {
      my $c = shift;
      if ($c->is_trusted_source || $c->is_trusted_source('1.2.3.4')) {
        ...
      }
    }

Validate if an IP address is in the ["trusted\_sources"](#trusted_sources) list. If no argument is
provided, then this helper will first check `tx->remote_proxy_address`
then `tx->remote_address`. Returns `1` if in the ["trusted\_sources"](#trusted_sources)
list, `0` if not, or `undef` if the IP address is invalid.

## process\_ip\_headers

    # From Controller context
    sub get_page {
      my $c = shift;
      my $matched_header = $c->process_ip_headers(1);
    }

Finds the first matching header from ["ip\_headers"](#ip_headers) and sets
`tx->remote_address` to the value (if a valid IP address), sets
`tx->remote_proxy_address` to the IP address of the upstream proxy, and
returns a hash of the name and value of the matched header. Otherwise returns
`0` if no match is found.

If any "truthy" value is passed as a parameter to this helper, it will first
run the ["is\_trusted\_source"](#is_trusted_source) helper (no arguments passed) and will return
`undef` if it returns a false value.

**WARNING!** Calling this helper when ["enabled"](#enabled) is set to true will likely
produce unexpected results.

## process\_scheme\_headers

    # From Controller context
    sub get_page {
      my $c = shift;
      my $matched_header = $c->process_scheme_headers(1);
    }

Finds the first matching header from ["scheme\_headers"](#scheme_headers) and sets
`req->url->base->scheme` to `https` if the header value matches any
defined in ["https\_values"](#https_values), and returns a hash of the name and value of the
matched header. Otherwise returns `0` if no match is found.

If any "truthy" value is passed as a parameter to this helper, it will first
run the ["is\_trusted\_source"](#is_trusted_source) helper (no arguments passed) and will return
`undef` if it returns a false value.

**WARNING!** Calling this helper when ["enabled"](#enabled) is set to true will likely
produce unexpected results.

## process\_protocol\_headers

Alias for ["process\_scheme\_headers"](#process_scheme_headers).

## process\_rfc7239\_header

    # From Controller context
    sub get_page {
      my $c = shift;
      my @matched_params = $c->process_rfc7239_header(1);
    }

Process an RFC 7239 ("Forwarded") HTTP header if found. See ["parse\_rfc7239"](#parse_rfc7239)
for more details. If the "Forwarded" header is found, this helper will return
an array of hashes of the matched RFC 7239 parameters and values, or `0` if no
matches are found.

If any "truthy" value is passed as a parameter to this helper, it will first
run the ["is\_trusted\_source"](#is_trusted_source) helper (no arguments passed) and will return
`undef` if it returns a false value.

**WARNING!** Calling this helper when ["enabled"](#enabled) is set to true will likely
produce unexpected results.

## process\_forwarded\_header

Alias for ["process\_rfc7239\_header"](#process_rfc7239_header).

## hide\_upstream\_headers

    # From Controller context
    sub get_page {
      my $c = shift;
      $c->hide_headers(1);
    }

Remove all headers defined in ["ip\_headers"](#ip_headers), ["scheme\_headers"](#scheme_headers), and
`Forwarded` from the request.

If any "truthy" value is passed as a parameter to this helper, it will first
run the ["is\_trusted\_source"](#is_trusted_source) helper (no arguments passed) and will return
`undef` if it returns a false value.

# CDN AND CLOUD SUPPORT

[Mojolicious::Plugin::TrustedProxy](https://metacpan.org/pod/Mojolicious::Plugin::TrustedProxy) is compatible with assumedly all
third-party content delivery networks and cloud providers. Below is an
incomplete list of some of the most well-known providers and the recommended
[config](#config) values to use for them.

## Akamai

- ip\_headers

    Set ["ip\_headers"](#ip_headers) to `['true-client-ip']` (unless you set this to a different
    value) and enable True Client IP in the origin server behavior for your site
    property. Akamai also supports `['x-forwarded-for']`, which is enabled by
    default in [Mojolicious::Plugin::TrustedProxy](https://metacpan.org/pod/Mojolicious::Plugin::TrustedProxy).

- scheme\_headers

    There is no known way to pass this by default with Akamai. It may be possible
    to pass a custom header via a combination of a Site Property variable and a
    custom behavior that injects an outgoing request header based on that variable,
    but this has not been tested or confirmed.

- trusted\_sources

    This is only possible if you have the
    [Site Shield](https://www.akamai.com/us/en/products/security/site-shield.jsp)
    product from Akamai. If so, set ["trusted\_sources"](#trusted_sources) to the complete list of
    IPs provided in your Site Shield map.

## AWS

- ip\_headers

    The AWS Elastic Load Balancer uses `['x-forwarded-for']`, which is enabled by
    default in [Mojolicious::Plugin::TrustedProxy](https://metacpan.org/pod/Mojolicious::Plugin::TrustedProxy).

- scheme\_headers

    The AWS Elastic Load Balancer uses `['x-forwarded-proto']`, which is enabled
    by default in [Mojolicious::Plugin::TrustedProxy](https://metacpan.org/pod/Mojolicious::Plugin::TrustedProxy).

- trusted\_sources

    Depending on your setup, this could be one of the `172.x.x.x` IP addresses
    or ranges within your Virtual Private Cloud, the IP address(es) of your Elastic
    or Application Load Balancer, or could be the public IP ranges for your AWS
    region. Go to
    [https://docs.aws.amazon.com/general/latest/gr/aws-ip-ranges.html](https://docs.aws.amazon.com/general/latest/gr/aws-ip-ranges.html) for an
    updated list of AWS's IPv4 and IPv6 CIDR ranges.

## Cloudflare

- ip\_headers

    Set ["ip\_headers"](#ip_headers) to `['cf-connecting-ip']`, or `['true-client-ip']` if
    using an enterprise plan. Cloudflare also supports `['x-forwarded-for']`,
    which is enabled by default in [Mojolicious::Plugin::TrustedProxy](https://metacpan.org/pod/Mojolicious::Plugin::TrustedProxy).

- scheme\_headers

    Cloudflare uses the `x-forwarded-proto` header, which is enabled by default
    in [Mojolicious::Plugin::TrustedProxy](https://metacpan.org/pod/Mojolicious::Plugin::TrustedProxy).

- trusted\_sources

    Go to [https://www.cloudflare.com/ips/](https://www.cloudflare.com/ips/) for an updated list of Cloudflare's
    IPv4 and IPv6 CIDR ranges.

# SECURITY

Caution should be taken that you set only the ["CONFIG"](#config) values necessary for
your application in a most-common-first order, and that your upstream proxies
remove any headers you do not want passed through to your application.

For example, if you use Cloudflare and set ["ip\_headers"](#ip_headers) to
`['x-real-ip', 'cf-connecting-ip']` and did not configure Cloudflare to
remove `x-real-ip` headers from requests, an attacker could use this trick
your application into using whatever IP he or she defines due to being passed
through your trusted proxy and the `x-real-ip` header being the first to be
evaluated.

# AUTHOR

Kage &lt;kage _AT_ kage _DOT_ wtf>

# BUGS

Please report any bugs or feature requests on Github:
[https://github.com/Kage/Mojolicious-Plugin-TrustedProxy](https://github.com/Kage/Mojolicious-Plugin-TrustedProxy)

- Hostnames not supported

    Excluding the `host` parameter of RFC 7239, this plugin does not currently
    support hostnames or hostname resolution and there are no plans to implement
    this. If you have a use case that requires this, please feel free to submit a
    pull request.

- HTTP 'Forwarded' only partially supported

    Only partial support for RFC 7239 is currently implemented, but this should
    work with most common use cases. The full specification allows for complex
    structures and quoting that is difficult to implement safely. Full RFC support
    is expected to be implemented soon.

# SEE ALSO

[Mojolicious::Plugin::RemoteAddr](https://metacpan.org/pod/Mojolicious::Plugin::RemoteAddr), [Mojolicious::Plugin::ClientIP::Pluggable](https://metacpan.org/pod/Mojolicious::Plugin::ClientIP::Pluggable),
[Mojolicious](https://metacpan.org/pod/Mojolicious), [Mojolicious::Guides](https://metacpan.org/pod/Mojolicious::Guides), [http://mojolicio.us](http://mojolicio.us).

# COPYRIGHT

MIT License

Copyright (c) 2019 Kage

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


Changes
0.05
    Added several helpers to make this plugin even more customizable
    Added minimum versions for some module dependencies

0.04
    Added support for "host" parameter of RFC-7239
    Changed default header order for better security

0.03
    Added support for RFC-7239 (HTTP Forwarded header)
    Brought all development configs up to date

0.02
    Renamed plugin to TrustedProxy
    Added a series of unit tests
    Added configurable HTTP header check values

0.01
    First release.


t/lib/TestApp.pm
package TestApp;
use Mojo::Base 'Mojolicious';

sub startup {
  my $self   = shift;
  my $r      = $self->routes;
  my $config = $self->config->{trustedproxy} || {};

  $self->plugin('TrustedProxy' => $config);

  # Returns current value of tx->remote_proxy_address
  $r->get(
    '/proxyip' => sub {
      my $c = shift;
      $c->render(text => $c->tx->remote_proxy_address || '');
    }
  );

  # Returns current value of tx->remote_address
  $r->get(
    '/ip' => sub {
      my $c = shift;
      $c->render(text => $c->tx->remote_address);
    }
  );

  # Returns current connection scheme as 'http' or 'https'
  $r->get(
    '/scheme' => sub {
      my $c = shift;
      $c->render(text => $c->req->is_secure ? 'https' : 'http');
    }
  );

  # Returns all header names
  $r->get(
    '/headers' => sub {
      my $c = shift;
      $c->render(json => $c->req->headers->names);
    }
  );
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Mojolicious plugin to set the remote address, connection scheme, and more from trusted upstream proxies",
  "AUTHOR" => "Kage <kage\@kage.wtf>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Mojolicious-Plugin-TrustedProxy",
  "LICENSE" => "mit",
  "NAME" => "Mojolicious::Plugin::TrustedProxy",
  "PREREQ_PM" => {
    "Data::Validate::IP" => "0.21",
    "Mojolicious" => "7.19",
    "Net::CIDR::Lite" => 0,
    "Net::IP::Lite" => 0
  },
  "TEST_REQUIRES" => {
    "Devel::Cover" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "lib::relative" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Validate::IP" => "0.21",
  "Devel::Cover" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Mojolicious" => "7.19",
  "Net::CIDR::Lite" => 0,
  "Net::IP::Lite" => 0,
  "Test::More" => 0,
  "lib::relative" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Mojolicious/Plugin/HeadsUp.pm
package Mojolicious::Plugin::HeadsUp;
use Mojo::Base 'Mojolicious::Plugin';
use Net::CIDR::Lite;
use Net::IP::Lite qw(ip_validate ip_is_ipv6ipv4 ip_transform);

# https://github.com/Kage/Mojolicious-Plugin-HeadsUp

our $VERSION = '0.02';

use constant DEBUG => $ENV{MOJO_HEADSUP_DEBUG} || 0;

sub register {
  my ($self, $app, $conf) = @_;

  $app->log->debug(sprintf('[%s] VERSION = %s', __PACKAGE__, $VERSION))
    if DEBUG;

  # Set config defaults if undefined
# TODO
#  $conf->{support_rfc7239} //= 1;

  $conf->{ip_headers}      //= ['x-real-ip', 'x-forwarded-for'];
  $conf->{ip_headers}        = [$conf->{ip_headers}]
    unless ref($conf->{ip_headers}) eq 'ARRAY';

  $conf->{scheme_headers}  //= ['x-ssl', 'x-forwarded-proto'];
  $conf->{scheme_headers}    = [$conf->{scheme_headers}]
    unless ref($conf->{scheme_headers}) eq 'ARRAY';

  $conf->{https_values}    //= ['1', 'true', 'https', 'on', 'enable', 'enabled'];
  $conf->{https_values}      = [$conf->{https_values}]
    unless ref($conf->{https_values}) eq 'ARRAY';

  $conf->{trusted_sources} //= ['127.0.0.0/8', '10.0.0.0/8'];
  $conf->{trusted_sources}   = [$conf->{trusted_sources}]
    unless ref($conf->{trusted_sources}) eq 'ARRAY';

  $conf->{hide_headers}    //= 0;

  # Assemble trusted source CIDR map
  my $cidr = Net::CIDR::Lite->new;
  foreach my $trust (@{$conf->{trusted_sources}}) {
    if (ref($trust) eq 'ARRAY') {
      $cidr->add_any(@$trust);
    } elsif (ref($trust) eq 'HASH') {
      $cidr->add_any(values(%$trust));
    } else {
      $cidr->add_any($trust);
    }
    $cidr->clean;
  }
  $app->defaults(
    'headsup.conf' => $conf,
    'headsup.cidr' => $cidr,
  );

  # Register helper
  $app->helper(is_trusted_source => sub {
    my $c = shift;
    my $ip = shift || $c->tx->original_remote_address || $c->tx->remote_address;
    my $cidr = $c->stash('headsup.cidr');
    return undef unless
      ip_validate($ip) && $cidr && $cidr->isa('Net::CIDR::Lite');
    $ip = ip_transform($ip, {convert_to => 'ipv4'}) if (ip_is_ipv6ipv4($ip));
    $c->app->log->debug(sprintf(
      '[%s] Testing if IP address "%s" is in trusted sources list',
      __PACKAGE__, $ip)) if DEBUG;
    return $cidr->find($ip);
  });

  # Register hook
  $app->hook(around_dispatch => sub {
    my ($next, $c) = @_;
    my $conf = $c->stash('headsup.conf');
    return $next->() unless defined $conf;

    # Validate that the upstream source IP is within the CIDR map
    my $src_addr = $c->tx->remote_address;
    unless (defined $src_addr && $c->is_trusted_source($src_addr)) {
      $c->app->log->debug(sprintf(
        '[%s] %s not found in trusted_sources CIDR map',
        __PACKAGE__, $src_addr)) if DEBUG;
      return $next->();
    }

    # Set forwarded IP address from header
    foreach my $header (@{$conf->{ip_headers}}) {
      if (my $ip = $c->req->headers->header($header)) {
        $c->app->log->debug(sprintf(
          '[%s] Matched on IP header "%s" (value: "%s")',
          __PACKAGE__, $header, $ip)) if DEBUG;
        $c->tx->original_remote_address($src_addr);
        $c->tx->remote_address($ip);
        last;
      }
    }

    # Set forwarded scheme from header
    foreach my $header (@{$conf->{scheme_headers}}) {
      if (my $scheme = $c->req->headers->header($header)) {
        if (!!$scheme && grep { lc $scheme eq $_ } @{$conf->{https_values}}) {
          $c->app->log->debug(sprintf(
            '[%s] Matched on HTTPS header "%s" (value: "%s")',
            __PACKAGE__, $header, $scheme)) if DEBUG;
          $c->req->url->base->scheme('https');
          last;
        }
      }
    }

    # Hide headers from the rest of the application
    if (!!$conf->{hide_headers}) {
      $c->app->log->debug(sprintf(
        '[%s] Removing headers from request', __PACKAGE__)) if DEBUG;
      $c->req->headers->remove($_) foreach @{$conf->{ip_headers}};
      $c->req->headers->remove($_) foreach @{$conf->{scheme_headers}};
    }

    # Carry on :)
    $next->();
  });

}

# For Desc:

#This plugin also supports parsing L<RFC 7239|http://tools.ietf.org/html/rfc7239>
#compliant C<Forwarded> headers and forwarded headers from
#L<Cloudflare|https://cloudflare.com>.

# For Config:

#=head2 support_rfc7239
#
#Enable support for parsing L<RFC 7239|http://tools.ietf.org/html/rfc7239>
#compliant C<Forwarded> HTTP headers. Default is C<1> (enabled).
#
#B<Note!> If enabled, the headers defined in L</ip_headers> and
#L</scheme_headers> will be ignored if the C<Forwarded> header is found.

1;
__END__
=head1 NAME

Mojolicious::Plugin::HeadsUp - Override request values using supported
headers from trusted upstream sources

=head1 VERSION

Version 0.02

=head1 SYNOPSIS

  use Mojolicious::Lite;

  plugin 'HeadsUp' => {
    ip_headers      => ['x-real-ip', 'x-forwarded-for'],
    scheme_headers  => ['x-ssl', 'x-forwarded-proto'],
    https_values    => ['1', 'true', 'https', 'on', 'enable', 'enabled'],
    trusted_sources => ['127.0.0.0/8', '10.0.0.0/8'],
    hide_headers    => 0,
  };

  get '/test' => sub {
    my $c = shift;
    $c->render(json => {
      'tx.remote_address'            => $c->tx->remote_address,
      'tx.original_remote_address'   => $c->tx->original_remote_address,
      'req.url.base.scheme'          => $c->req->url->base->scheme,
      'is_trusted_source'            => $c->is_trusted_source,
      'is_trusted_source("1.1.1.1")' => $c->is_trusted_source('1.1.1.1'),
    });
  };

  app->start;

=head1 DESCRIPTION

L<Mojolicious::Plugin::HeadsUp> modifies every L<Mojolicious> request transaction
to override connecting user agent values only when the request comes from trusted
upstream sources. You can specify multiple request headers where trusted upstream
sources define the real user agent IP address or the real connection scheme, or
disable either, and can hide the headers from the rest of the application if
needed.

Debug logging can be enabled by setting the C<MOJO_HEADSUP_DEBUG> environment
variable.

Build status:

=begin html

<a href="https://travis-ci.org/Kage/Mojolicious-Plugin-HeadsUp">
<img src="https://travis-ci.org/Kage/Mojolicious-Plugin-HeadsUp.svg?branch=master">
</a>

=end html

=head1 CONFIG

=head2 ip_headers

List of zero, one, or many HTTP headers where the real user agent IP address
will be defined by the trusted upstream sources. The first matched header is
used. An empty value will disable this and keep the original scheme value.
Default is C<['x-real-ip', 'x-forwarded-for']>.

If a header is matched in the request, then C<< tx->remote_address >> is set to
the value, and C<< tx->original_remote_address >> is set to the IP address of the
upstream source.

=head2 scheme_headers

List of zero, one, or many HTTP headers where the real user agent connection
scheme will be defined by the trusted upstream sources. The first matched header
is used. An empty value will disable this and keep the original remote address
value. Default is C<['x-ssl', 'x-forwarded-proto']>.

This tests that the header value is "truthy" but does not contain the literal
barewords C<http>, C<off>, or C<false>. If the header contains any other
"truthy" value, then C<< req->url->base->scheme >> is set to C<https>.

=head2 https_values

List of values to consider as "truthy" when evaluating the headers in
L</scheme_headers>. Default is
C<['1', 'true', 'https', 'on', 'enable', 'enabled']>.

=head2 trusted_sources

List of one or more IP addresses or CIDR classes that are trusted upstream
sources. (B<Warning!> An empty value will trust from all IPv4 sources!) Default
is C<['127.0.0.0/8', '10.0.0.0/8']>.

Supports all IP, CIDR, and range definition types from L<Net::CIDR::Lite>.

=head2 hide_headers

Hide all headers defined in L</ip_headers> and L</scheme_headers> from the rest
of the application when coming from trusted upstream sources. Default is C<0>
(disabled).

=head1 HELPERS

=head2 is_trusted_source

  # From Controller context
  sub get_page {
    my $c = shift;
    if ($c->is_trusted_source || $c->is_trusted_source('1.2.3.4')) {
      ...
    }
  }

Validate if an IP address is in the L</trusted_sources> list. If no argument is
provided, then this helper will first check C<< tx->original_remote_address >>
then C<< tx->remote_address >>. Returns C<1> if in the L</trusted_sources> list,
C<0> if not, or C<undef> if the IP address is invalid.

=head1 CDN AND CLOUD SUPPORT

L<Mojolicious::Plugin::HeadsUp> is compatible with assumedly all third-party
content delivery networks and cloud providers. Below is an incomplete list of
some of the most well-known providers and the recommended config values to use
for them.

=head2 Akamai

=over

=item ip_headers

Set L</ip_headers> to C<['true-client-ip']> (unless you set this to a different
value) and enable True Client IP in the origin server behavior for your site
property. Akamai also supports C<['x-forwarded-for']>, which is enabled by
default in L<Mojolicious::Plugin::HeadsUp>.

=item scheme_headers

There is no known way to pass this by default with Akamai. It may be possible
to pass a custom header via a combination of a site property variable and a
custom rule that injects an outgoing request header based on that variable,
but this has not been tested.

=item trusted_sources

This is only possible if you have the
L<Site Shield|https://www.akamai.com/us/en/products/security/site-shield.jsp>
product from Akamai. If so, set L</trusted_sources> to the complete list of
IPs provided in your Site Shield map.

=back

=head2 AWS

=over

=item ip_headers

The AWS Elastic Load Balancer uses C<['x-forwarded-for']>, which is enabled by
default in L<Mojolicious::Plugin::HeadsUp>.

=item scheme_headers

The AWS Elastic Load Balancer uses C<['x-forwarded-proto']>, which is enabled
by default in L<Mojolicious::Plugin::HeadsUp>.

=item trusted_sources

Depending on your setup, this may be one of the C<172.x.x.x> IP addresses within
your virtual private cloud, or may be within the public IP ranges for your AWS
region. Go to
L<https://docs.aws.amazon.com/general/latest/gr/aws-ip-ranges.html> for an
updated list of AWS's IPv4 and IPv6 CIDR ranges.

=back

=head2 Cloudflare

=over

=item ip_headers

Set L</ip_headers> to C<['cf-connecting-ip']>, or C<['true-client-ip']> if
using an enterprise plan. Cloudflare also supports C<['x-forwarded-for']>,
which is enabled by default in L<Mojolicious::Plugin::HeadsUp>.

=item scheme_headers

Cloudflare uses the C<x-forwarded-proto> header, which is enabled by default
in L<Mojolicious::Plugin::HeadsUp>.

=item trusted_sources

Go to L<https://www.cloudflare.com/ips/> for an updated list of Cloudflare's
IPv4 and IPv6 CIDR ranges.

=back

=head1 AUTHOR

Kage <kage@kage.wtf>

=head1 BUGS

Please report any bugs or feature requests on Github:
L<https://github.com/Kage/Mojolicious-Plugin-HeadsUp>

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<Mojolicious::Plugin::RemoteAddr>,
L<Net::CIDR::Lite>, L<http://mojolicio.us>.

=head1 COPYRIGHT

MIT License

Copyright (c) 2019 Kage

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

=cut
